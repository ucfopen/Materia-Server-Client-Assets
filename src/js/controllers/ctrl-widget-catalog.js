import toSnakeCase from 'js-snakecase'

const app = angular.module('materia')

// when the current url matches the widget catalog
// then use html5Mode to allow the filters and search to update the url
// and add ng-animate
// ex: /widgets or /widgets/1-adventure
if (window.location.href.match(/\/widgets($|\/\D|\?)/g)) {
	app.config(function ($locationProvider) {
		$locationProvider.html5Mode({
			enabled: true,
			requireBase: false,
		})
	})

	app.requires.push('ngAnimate')
}

app.controller('WidgetCatalogCtrl', function (Please, $scope, $window, $location, WidgetSrv) {
	const filterList = {} // key is raw filter name, value is filter object
	const mapCleanToFilter = {} // key is clean name, value is filter object
	const displayedWidgets = []
	let allWidgets = []

	const resetFilters = () => {
		for (i in filterList) delete filterList[i]
		for (i in mapCleanToFilter) delete mapCleanToFilter[i]
	}

	const showFilters = () => {
		$scope.isShowingFilters = true
	}

	const clearFilters = (hideFilters = true) => {
		// set all filters to false
		for (let filter in filterList) {
			setFilter(filterList[filter], false)
		}

		if (hideFilters) $scope.isShowingFilters = false
		_updateWidgetDisplay()
	}

	const clearFiltersAndSearch = (hideFilters = true) => {
		$scope.search = ''
		clearFilters(hideFilters)
	}

	const toggleFilter = (filterName) => {
		const filter = filterList[filterName]
		setFilter(filter, !filter.isActive)
		_updateWidgetDisplay()
	}

	const setFilter = (filter, value) => {
		filter.isActive = value
		const cleanName = filter.clean
		const val = filter.isActive || null
		$location.search(cleanName, val).replace()
	}

	const _registerFilter = (name, appendLabel = '') => {
		const clean = toSnakeCase(name)
		const filter = {
			isActive: false,
			text: `${name}${appendLabel}`,
			clean,
		}
		filterList[name] = filter
		mapCleanToFilter[clean] = filter
	}

	const _getFiltersFromWidgets = (widgets) => {
		widgets.forEach((widget) => {
			if (widget.meta_data.hasOwnProperty('features')) {
				widget.meta_data.features.forEach((feature) => {
					if (!filterList.hasOwnProperty(feature)) _registerFilter(feature)
				})
			}

			if (widget.meta_data.hasOwnProperty('supported_data')) {
				widget.meta_data.supported_data.forEach((data) => {
					if (!filterList.hasOwnProperty(data)) _registerFilter(data, ' Questions')
				})
			}

			if (widget.meta_data.hasOwnProperty('accessibility_options')) {
				widget.meta_data.accessibility_options.forEach((option, index) => {
					if (!filterList.hasOwnProperty(option) && index == 0)
						_registerFilter('Keyboard Accessible')
					if (!filterList.hasOwnProperty(option) && index == 1)
						_registerFilter('Screen Reader Accessible')
				})
			}
		})
	}

	// checks filters and returns true if the widget should be shown, false if filtered out
	const _isWidgetVisible = (widget) => {
		const wFeatures = widget.meta_data.features
		const wSupport = widget.meta_data.supported_data
		const wAccessibility = _accessOptionsToFilters(widget.meta_data.accessibility_options)

		// check for filter matches
		for (let filterName in filterList) {
			// this filter is active
			if (filterList[filterName].isActive) {
				// widget has features/support/accessibility and the feature isn't in either
				if (
					(!wFeatures || !wFeatures.includes(filterName)) &&
					(!wSupport || !wSupport.includes(filterName)) &&
					(!wAccessibility || !wAccessibility.includes(filterName))
				) {
					return false
				}
			}
		}

		// check for search matches
		if ($scope.search.length) {
			const re = new RegExp($scope.search, 'i')
			return re.test(widget.name)
		}

		return true
	}

	const _updateWidgetDisplay = () => {
		const widgets = allWidgets.filter((w) => _isWidgetVisible(w))
		$scope.count = widgets.length
		$scope.activeFilters = Object.keys(filterList).filter((key) => filterList[key].isActive)
		$scope.isFiltered = widgets.length != allWidgets.length

		if ($scope.isFiltered) {
			// don't display featured - place everything in widgets
			$scope.widgets = widgets
		} else {
			// no filters active - show the featured list
			$scope.widgets = widgets.filter((w) => w.in_catalog != '1')
		}

		Please.$apply()
	}

	const _onSearch = () => {
		const val = $scope.search || null
		$location.search('search', val).replace()
		_updateWidgetDisplay()
	}

	const _loadWidgets = () => {
		// load list of widgets
		WidgetSrv.getWidgetsByType('all').then((loaded) => {
			if (!loaded || !loaded.length || !loaded.length > 0) {
				resetFilters()
				loaded = []
			}
			allWidgets = loaded

			allWidgets.forEach(_getAccessibilityOptions)

			$scope.noWidgetsInstalled = allWidgets.length == 0

			_getFiltersFromWidgets(allWidgets)

			// memoize icon paths
			allWidgets.forEach((widget) => {
				widget.icon = Materia.Image.iconUrl(widget.dir, 275)
			})

			$scope.totalWidgets = allWidgets.length
			$scope.featuredWidgets = allWidgets.filter((w) => w.in_catalog == '1')

			// start watching search input
			$scope.$watch('search', _onSearch)

			// load the filters now because they come from the widgets
			_getFiltersFromURL()

			_updateWidgetDisplay()

			// prevents animation on initial load; set after initial $apply
			$scope.ready = true
		})
	}

	const _getFiltersFromURL = () => {
		$scope.isShowingFilters = false
		for (let key in $location.search()) {
			if (key == 'search') {
				$scope.search = $location.search().search
			} else if (mapCleanToFilter[key]) {
				mapCleanToFilter[key].isActive = true
				$scope.isShowingFilters = true
			}
		}
	}

	// Makes sure accessibility rating is valid
	const _getAccessibilityOptions = (widget, index) => {
		const valid_options = ['none', 'limited', 'full']

		if (widget.meta_data.hasOwnProperty('accessibility_options')) {
			widget.meta_data.accessibility_options.forEach((val, index, arr) => {
				if (!valid_options.includes(val.toLowerCase())) {
					arr[index] = 'None'
				}
			})
		}
	}

	const _accessOptionsToFilters = (data) => {
		const valid_options = ['limited', 'full']
		let valid_filters = []

		if (data == undefined) return valid_filters

		for (let i = 0; i < 2; i++) {
			if (i >= data.length) break

			if (valid_options.includes(data[i].toLowerCase())) {
				valid_filters.push(i == 0 ? 'Keyboard Accessible' : 'Screen Reader Accessible')
			}
		}

		return valid_filters
	}

	$scope.search = ''
	$scope.totalWidgets = -1
	$scope.count = -1
	$scope.noWidgetsInstalled = false
	$scope.isShowingFilters = false
	$scope.ready = false
	$scope.activeFilters = []
	$scope.showFilters = showFilters
	$scope.clearFilters = clearFilters
	$scope.clearFiltersAndSearch = clearFiltersAndSearch
	$scope.featuredWidgets = []
	$scope.toggleFilter = toggleFilter
	$scope.widgets = []
	$scope.filters = filterList
	$scope.isFiltered = false

	// with html mode on, angular processes location changes
	// We have to manually change url when needed
	$scope.$on('$locationChangeStart', (e, newUrl) => {
		if (!newUrl.match(/\/widgets($|\/\D|\?)/g)) {
			$window.location = newUrl
		} else if ($scope.totalWidgets != -1) {
			// handles the "Widget Catalog" link in the header
			const urlParamCount = Object.keys($location.search()).length
			if (urlParamCount == 0) {
				clearFiltersAndSearch(false)
			}
		}
	})

	_loadWidgets()

	/* develblock:start */
	// these method are exposed for testing
	$scope.jestTest = {
		getLocalVar: (name) => eval(name),
	}
	/* develblock:end */
})
