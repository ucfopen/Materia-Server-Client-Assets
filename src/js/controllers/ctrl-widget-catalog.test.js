describe('widgetCatalogController', () => {
	let $controller
	let $q
	let $scope
	let $location
	let location
	let mockIconUrl
	let mockWidgetSrv
	let mockLocationSearch
	let widgetsToReturn

	const mockUrl = 'http://localhost/widgets?customizable&search=widget&fake_feature'
	const widget1 = {
		id: 1,
		dir: 'mockDir1',
		in_catalog: '1',
		meta_data: {
			about: 'information about the widget',
			demo: '1',
			excerpt: 'more information about the widget',
			features: ['feature1', 'feature3'],
			supported_data: ['supported1', 'supported_three', 'SuPpOrTeD FoUr!!'],
			benefits: ['https://www.youtube.com/watch?v=dQw4w9WgXcQ']
		},
		name: 'widget1'
	}
	const widget2 = {
		id: 2,
		dir: 'mockDir2',
		in_catalog: '0',
		meta_data: {
			about: 'information about the widget',
			demo: '2',
			excerpt: 'more information about the widget',
			features: ['feature2', 'feature3'],
			supported_data: ['supported1', 'supported2']
		},
		name: 'widget2'
	}
	const widget3 = {
		id: 3,
		dir: 'mockDir3',
		in_catalog: '0',
		meta_data: {
			about: 'information about the widget',
			demo: '3',
			excerpt: 'more information about the widget',
			features: [],
			supported_data: []
		},
		name: 'widget3'
	}

	const widgetWithoutFeatures = {
		id: 4,
		dir: 'mockDir4',
		in_catalog: '0',
		meta_data: {
			about: 'information about the widget',
			demo: '4',
			excerpt: 'more information about the widget'
			// features and supported_data are null!!
		},
		name: 'widgetWithoutFeatures'
	}

	beforeEach(() => {
		widgetsToReturn = [widget1, widget2, widget3, widgetWithoutFeatures]
		// mocks getWidgetsByType promise with a synchronous implementation
		// why? because it's an internal private method and getting a handle
		// to wait for it to finish is difficult
		let getWidgetsByTypeImmediately = jest.fn().mockReturnValue({
			then: cb => {
				cb(widgetsToReturn)
			}
		})
		// mock all the required services
		mockWidgetSrv = { getWidgetsByType: getWidgetsByTypeImmediately }
		let app = angular.module('materia')
		app.factory('Please', () => ({ $apply: jest.fn() }))
		app.factory('selectedWidgetSrv', () => ({}))
		app.factory('dateTimeServ', () => ({}))
		app.factory('widgetSrv', () => mockWidgetSrv)

		// mock Materia.Image.iconUrl
		mockIconUrl = jest.fn().mockReturnValue('widget.jpg')
		window.Materia = { Image: { iconUrl: mockIconUrl } }

		// mock window.location
		let mockWindow = {}
		let mockLocationSet = jest.fn(l => (location = l))
		let mockLocationGet = jest.fn(() => location)
		Object.defineProperty(mockWindow, 'location', {
			get: mockLocationGet,
			set: mockLocationSet
		})
		app.factory('$window', () => mockWindow)

		// manually set the url
		window.history.pushState({}, '', mockUrl)

		// build a mock $scope
		$scope = {
			$watch: jest.fn(),
			$on: jest.fn()
		}

		mockLocationSearch = { search: '' }

		require('angular-animate')
		require('./ctrl-widget-catalog')

		inject((_$controller_, _$location_, _$q_) => {
			$controller = _$controller_
			$location = _$location_
			$q = _$q_

			// mock to get/set url params
			$location.search = jest.fn((key, val) => {
				if (!key) {
					return mockLocationSearch
				}
				return {
					replace: jest.fn()
				}
			})
		})
	})

	it('defines expected scope vars', () => {
		$controller('widgetCatalogCtrl', { $scope })
		expect(Object.keys($scope)).toMatchSnapshot()
	})

	it('loads widgets from the widget service', () => {
		$controller('widgetCatalogCtrl', { $scope })
		expect(mockWidgetSrv.getWidgetsByType).toHaveBeenCalledTimes(1)
		expect(mockWidgetSrv.getWidgetsByType).toHaveBeenCalledWith('all')
	})

	it('uses Materia.Image.iconUrl to get each widget icon', () => {
		$controller('widgetCatalogCtrl', { $scope })
		expect(mockIconUrl).toHaveBeenCalledTimes(4)
		expect(mockIconUrl).toHaveBeenCalledWith('mockDir1', 275)
		expect(mockIconUrl).toHaveBeenCalledWith('mockDir2', 275)
		expect(mockIconUrl).toHaveBeenCalledWith('mockDir3', 275)
		expect(mockIconUrl).toHaveBeenCalledWith('mockDir4', 275)
	})

	it('handles no widgets', () => {
		widgetsToReturn = []
		$controller('widgetCatalogCtrl', { $scope })

		expect($scope.widgets).toHaveLength(0)
		expect($scope.featuredWidgets).toHaveLength(0)
		expect($scope.count).toEqual(0)
		expect($scope.noWidgetsInstalled).toBe(true)
		expect($scope.isFiltered).toBe(false)
		expect($scope.activeFilters).toHaveLength(0)
	})

	it('initializes with no filters and search on initial load', () => {
		$controller('widgetCatalogCtrl', { $scope })

		expect(Object.keys($scope)).toMatchSnapshot()
		expect($scope.search).toBe('')
		expect($scope.activeFilters).toEqual([])
		expect($scope.isShowingFilters).toBe(false)
		expect($scope.isFiltered).toBe(false)

		expect($scope.count).toBe(4)

		expect($scope.featuredWidgets).toHaveLength(1)
		expect($scope.featuredWidgets).toContain(widget1)

		expect($scope.widgets).toHaveLength(3)
		expect($scope.widgets).toContain(widget2)
		expect($scope.widgets).toContain(widget3)
		expect($scope.widgets).toContain(widgetWithoutFeatures)

		expect($scope.filters).toMatchSnapshot()
	})

	it('implements url based filters and search on initial load', () => {
		mockLocationSearch = {
			feature1: true,
			search: 'widget',
			invalid_feature: true,
			supported_three: true
		}
		$controller('widgetCatalogCtrl', { $scope })

		expect(Object.keys($scope)).toMatchSnapshot()
		expect($scope.search).toBe('widget')
		expect($scope.activeFilters).toEqual(['feature1', 'supported_three'])
		expect($scope.isShowingFilters).toBe(true)
		expect($scope.isFiltered).toBe(true)

		expect($scope.count).toBe(1)

		// note: featuredWidgets is not altered by filters or search
		expect($scope.featuredWidgets).toHaveLength(1)
		expect($scope.featuredWidgets).toContain(widget1)

		// only one widget matches the search and filter options
		expect($scope.widgets).toHaveLength(1)
		expect($scope.widgets).toContain(widget1)

		expect($scope.filters).toMatchSnapshot()
	})

	it('properly generates clean filter names', () => {
		$controller('widgetCatalogCtrl', { $scope })
		const mapCleanToFilter = $scope.jestTest.getLocalVar('mapCleanToFilter')
		expect(Object.keys(mapCleanToFilter)).toMatchSnapshot()
	})

	it('toggling on a filter updates scope', () => {
		$controller('widgetCatalogCtrl', { $scope })
		expect($scope.isFiltered).toBe(false)
		expect($scope.activeFilters).toHaveLength(0)
		expect($scope.filters['feature1'].isActive).toBe(false)
		expect($scope.widgets).toHaveLength(3)
		expect($scope.count).toBe(4)

		$scope.toggleFilter('feature1') // toggle on

		expect($scope.isFiltered).toBe(true)
		expect($scope.activeFilters).toHaveLength(1)
		expect($scope.filters['feature1'].isActive).toBe(true)
		expect($scope.widgets).toHaveLength(1)
		expect($scope.widgets).toMatchSnapshot()
		expect($scope.count).toBe(1)
	})

	it('will filter widgets based on a search query', () => {
		$controller('widgetCatalogCtrl', { $scope })
		const _onSearch = $scope.jestTest.getLocalVar('_onSearch')

		expect($scope.count).toBe(4)

		// search for widget 4
		$scope.search = 'widgetWithoutFeatures'
		_onSearch()

		expect($scope.count).toBe(1)
		expect($scope.widgets).toHaveLength(1)
		expect($scope.widgets).toContain(widgetWithoutFeatures)

		// search for widget 3
		$scope.search = 'widget3'
		_onSearch()

		expect($scope.count).toBe(1)
		expect($scope.widgets).toHaveLength(1)
		expect($scope.widgets).toContain(widget3)
	})

	it('can toggle whether the filters are showing', () => {
		$controller('widgetCatalogCtrl', { $scope })
		$scope.showFilters()
		expect($scope.isShowingFilters).toBe(true)
		$scope.clearFilters()
		expect($scope.isShowingFilters).toBe(false)
	})

	it('can clear filters and search', () => {
		$controller('widgetCatalogCtrl', { $scope })
		const _onSearch = $scope.jestTest.getLocalVar('_onSearch')
		$scope.search = 'widget1'
		_onSearch() // enable search
		$scope.toggleFilter('feature1') // toggle on
		expect($scope.activeFilters).toHaveLength(1)
		expect($scope.search).toBe('widget1')
		expect($scope.count).toBe(1)
		expect($scope.widgets).toHaveLength(1)

		// clear everything
		$scope.clearFiltersAndSearch()

		expect($scope.activeFilters).toHaveLength(0)
		expect($scope.search).toBe('')
		expect($scope.count).toBe(4)
		expect($scope.widgets).toHaveLength(3)
	})

	it('handles no widgets having features', () => {
		widgetsToReturn = [widgetWithoutFeatures]
		$controller('widgetCatalogCtrl', { $scope })
		// $scope.toggleFilter('feature1') // toggle on
		expect($scope.count).toEqual(1)
	})

	it('omits widgets with null features when a feature filter is enabled', () => {
		widgetsToReturn = [widget1, widgetWithoutFeatures]
		$controller('widgetCatalogCtrl', { $scope })
		expect($scope.count).toEqual(2)
		$scope.toggleFilter('feature1') // toggle on
		expect($scope.count).toEqual(1)
	})
})
