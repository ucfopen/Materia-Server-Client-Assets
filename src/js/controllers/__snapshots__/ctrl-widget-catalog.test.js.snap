// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetCatalogCtrl defines expected scope vars 1`] = `
Array [
  "$watch",
  "$on",
  "search",
  "totalWidgets",
  "count",
  "noWidgetsInstalled",
  "isShowingFilters",
  "ready",
  "activeFilters",
  "showFilters",
  "clearFilters",
  "clearFiltersAndSearch",
  "featuredWidgets",
  "toggleFilter",
  "widgets",
  "filters",
  "isFiltered",
  "jestTest",
]
`;

exports[`WidgetCatalogCtrl implements url based filters and search on initial load 1`] = `
Array [
  "$watch",
  "$on",
  "search",
  "totalWidgets",
  "count",
  "noWidgetsInstalled",
  "isShowingFilters",
  "ready",
  "activeFilters",
  "showFilters",
  "clearFilters",
  "clearFiltersAndSearch",
  "featuredWidgets",
  "toggleFilter",
  "widgets",
  "filters",
  "isFiltered",
  "jestTest",
]
`;

exports[`WidgetCatalogCtrl implements url based filters and search on initial load 2`] = `
Object {
  "Keyboard Accessible": Object {
    "clean": "keyboard_accessible",
    "isActive": false,
    "text": "Keyboard Accessible",
  },
  "Screen Reader Accessible": Object {
    "clean": "screen_reader_accessible",
    "isActive": false,
    "text": "Screen Reader Accessible",
  },
  "SuPpOrTeD FoUr!!": Object {
    "clean": "su_pp_or_te_d_fo_ur",
    "isActive": false,
    "text": "SuPpOrTeD FoUr!! Questions",
  },
  "feature1": Object {
    "clean": "feature1",
    "isActive": true,
    "text": "feature1",
  },
  "feature2": Object {
    "clean": "feature2",
    "isActive": false,
    "text": "feature2",
  },
  "feature3": Object {
    "clean": "feature3",
    "isActive": false,
    "text": "feature3",
  },
  "supported1": Object {
    "clean": "supported1",
    "isActive": false,
    "text": "supported1 Questions",
  },
  "supported2": Object {
    "clean": "supported2",
    "isActive": false,
    "text": "supported2 Questions",
  },
  "supported_three": Object {
    "clean": "supported_three",
    "isActive": true,
    "text": "supported_three Questions",
  },
}
`;

exports[`WidgetCatalogCtrl initializes with no filters and search on initial load 1`] = `
Array [
  "$watch",
  "$on",
  "search",
  "totalWidgets",
  "count",
  "noWidgetsInstalled",
  "isShowingFilters",
  "ready",
  "activeFilters",
  "showFilters",
  "clearFilters",
  "clearFiltersAndSearch",
  "featuredWidgets",
  "toggleFilter",
  "widgets",
  "filters",
  "isFiltered",
  "jestTest",
]
`;

exports[`WidgetCatalogCtrl initializes with no filters and search on initial load 2`] = `
Object {
  "Keyboard Accessible": Object {
    "clean": "keyboard_accessible",
    "isActive": false,
    "text": "Keyboard Accessible",
  },
  "Screen Reader Accessible": Object {
    "clean": "screen_reader_accessible",
    "isActive": false,
    "text": "Screen Reader Accessible",
  },
  "SuPpOrTeD FoUr!!": Object {
    "clean": "su_pp_or_te_d_fo_ur",
    "isActive": false,
    "text": "SuPpOrTeD FoUr!! Questions",
  },
  "feature1": Object {
    "clean": "feature1",
    "isActive": false,
    "text": "feature1",
  },
  "feature2": Object {
    "clean": "feature2",
    "isActive": false,
    "text": "feature2",
  },
  "feature3": Object {
    "clean": "feature3",
    "isActive": false,
    "text": "feature3",
  },
  "supported1": Object {
    "clean": "supported1",
    "isActive": false,
    "text": "supported1 Questions",
  },
  "supported2": Object {
    "clean": "supported2",
    "isActive": false,
    "text": "supported2 Questions",
  },
  "supported_three": Object {
    "clean": "supported_three",
    "isActive": false,
    "text": "supported_three Questions",
  },
}
`;

exports[`WidgetCatalogCtrl properly generates clean filter names 1`] = `
Array [
  "feature1",
  "feature3",
  "supported1",
  "supported_three",
  "su_pp_or_te_d_fo_ur",
  "keyboard_accessible",
  "screen_reader_accessible",
  "feature2",
  "supported2",
]
`;

exports[`WidgetCatalogCtrl toggling on a filter updates scope 1`] = `
Array [
  Object {
    "dir": "mockDir1",
    "icon": "widget.jpg",
    "id": 1,
    "in_catalog": "1",
    "meta_data": Object {
      "about": "information about the widget",
      "accessibility_options": Array [
        "Full",
        "None",
      ],
      "demo": "1",
      "excerpt": "more information about the widget",
      "features": Array [
        "feature1",
        "feature3",
      ],
      "supported_data": Array [
        "supported1",
        "supported_three",
        "SuPpOrTeD FoUr!!",
      ],
    },
    "name": "widget1",
  },
]
`;
